### PR TITLE
P4 1789 add move start endpoint

### PR DESCRIPTION
**Pull request by martyn-w**
### Jira link

P4-1789

### What?

I have added/removed/altered:

- [x] adds /move/id/start endpoint to mark a move as in_transit

### Why?

I am doing this because:

- so moves can be started

### Have you? (optional)

- [X] Updated API docs if necessary	

### Deployment risks (optional)

- FE unaware of in_transit status and will not show them

### CODE REVIEW

File: app/controllers/api/move_events_controller.rb [hunk 1]
```diff
@@ -4,81 +4,72 @@ module Api
...
-    ACCEPT_PARAMS = [:type, attributes: %i[timestamp notes]].freeze
+    APPROVE_PARAMS = [:type, attributes: %i[timestamp date]].freeze
     CANCEL_PARAMS = [:type, attributes: %i[timestamp cancellation_reason cancellation_reason_comment notes]].freeze
-    COMPLETE_PARAMS = [:type, attributes: %i[timestamp notes]].freeze
     LOCKOUT_PARAMS = [:type, attributes: %i[timestamp notes], relationships: { from_location: {} }].freeze
     REDIRECT_PARAMS = [:type, attributes: %i[timestamp notes], relationships: { to_location: {} }].freeze
-    APPROVE_PARAMS = [:type, attributes: %i[timestamp date]].freeze
     REJECT_PARAMS = [:type, attributes: %i[timestamp rejection_reason cancellation_reason_comment rebook]].freeze
 
+    STANDARD_PARAMS = [:type, attributes: %i[timestamp notes]].freeze # for accept, complete and start move events
```
Comment (alexdesi) [inline app/controllers/api/move_events_controller.rb:13]: What about calling this `COMMON_PARAMS` or `SHARED_PARAMS`?

File: app/jobs/prepare_move_notifications_job.rb [hunk 2]
```diff
@@ -41,9 +41,9 @@ def build_notification_id(subscription, type_id, topic, action_name)
 
   def should_email?(move)
     # NB: only email for:
-    #   * move.status must be Requested, Booked or Cancelled (not Proposed or Completed) moves, AND
+    #   * move.status must be Requested, Booked, In_Transit or Cancelled (not Proposed or Completed) moves, AND
```
Comment (alexdesi) [inline app/jobs/prepare_move_notifications_job.rb:44]: is the case of `In_Transit` correct? (camel and underscore case)

File: app/services/event_log/move_runner.rb [hunk 3]
```diff
@@ -11,27 +11,29 @@ def call
...
         when Event::LOCKOUT
           # no action to perform when a move is locked out, this event is purely for auditing
         when Event::REDIRECT
           move.to_location = event.to_location
+        when Event::REJECT
+          move.status = Move::MOVE_STATUS_CANCELLED
+          move.rejection_reason = event.rejection_reason
+          move.cancellation_reason = Move::CANCELLATION_REASON_REJECTED
+          move.cancellation_reason_comment = event.cancellation_reason_comment
+          move.rebook if event.rebook?
```
Comment (alexdesi) [inline app/services/event_log/move_runner.rb:34]: can we rebook outside the REJECT event? 
or `Event::REJECT` must be only inside this when-branch?

File: app/services/move_events/params_validator.rb [hunk 4]
```diff
@@ -6,9 +6,10 @@ class ParamsValidator
 
     ACCEPTABLE_PLURAL_VERBS = %w[cancels completes approves rejects].freeze
 
-    attr_reader :timestamp, :type, :date, :cancellation_reason, :rejection_reason
+    attr_reader :timestamp, :type, :date, :cancellation_reason, :rejection_reason, :from_location, :to_location
 
-    validates :type, presence: true, inclusion: { in: %w[cancel complete lockouts redirects approve reject accepts] }
+    # TODO: if/when pluralising types, existing move event records will need migrating
```
Comment (alexdesi) [inline app/services/move_events/params_validator.rb:11]: sorry, is not clear to me what `pluralising types` mean, could you expand a bit on this? when do we need to do it?

File: spec/requests/api/move_events_controller_lockout_spec.rb [hunk 5]
```diff
@@ -108,18 +79,68 @@
...
+            },
+          }
+        end
+
+        it_behaves_like 'an endpoint that responds with error 422' do
+          let(:errors_422) do
+            [{
+              'title' => 'Invalid from_location',
+              'detail' => 'Validation failed: From location was not found',
+            }]
```
Comment (alexdesi) [inline spec/requests/api/move_events_controller_lockout_spec.rb:142]: we are testing 2 times the 422 in this spec (which is a bit expensive to do here),
maybe we can test this case elsewhere?

File: spec/requests/api/move_events_controller_redirect_spec.rb [hunk 6]
```diff
@@ -108,18 +79,68 @@
...
+              'detail' => 'Validation failed: Type is not included in the list',
+            }]
+          end
+        end
+      end
+
+      context 'with a missing to_location relationship' do
+        let(:redirect_params) { { data: { type: 'redirects', attributes: { timestamp: '2020-04-23T18:25:43.511Z' } } } }
+
+        it_behaves_like 'an endpoint that responds with error 422' do
```
Comment (alexdesi) [inline spec/requests/api/move_events_controller_redirect_spec.rb:116]: it sounds we should should remove validation tests from here

File: spec/requests/api/move_events_controller_redirect_spec.rb [hunk 7]
```diff
@@ -108,18 +79,68 @@
...
+          let(:errors_422) do
+            [{
+              'title' => 'Invalid to_location',
+              'detail' => 'Validation failed: To location is missing',
+            }]
+          end
+        end
+      end
+
+      context 'with a non-existent to_location' do
```
Comment (alexdesi) [inline spec/requests/api/move_events_controller_redirect_spec.rb:126]: This is another validation test we could move into proper validation tests.

File: spec/services/event_log/move_runner_spec.rb [hunk 8]
```diff
@@ -166,6 +166,30 @@
...
+    context 'when the move is already in_transit' do
+      let!(:move) { create(:move, :in_transit) }
+
+      it_behaves_like 'it does not call the Notifier'
+
+      it 'does not change the move status' do
+        expect { runner.call }.not_to change(move, :status).from('in_transit')
+      end
+    end
+  end
```
Comment (alexdesi) [inline spec/services/event_log/move_runner_spec.rb:191]: These are very clear 👍

File: app/controllers/api/move_events_controller.rb (hunk 1 above)
Comment (martyn-w) [reply app/controllers/api/move_events_controller.rb:13]: `COMMON_PARAMS` sounds good 👍  I'll rename this

File: app/jobs/prepare_move_notifications_job.rb (hunk 2 above)
Comment (martyn-w) [reply app/jobs/prepare_move_notifications_job.rb:44]: the actual name is `in_transit` - I'll rename them all in this comment

File: app/services/event_log/move_runner.rb (hunk 3 above)
Comment (martyn-w) [reply app/services/event_log/move_runner.rb:34]: hi @alexdesi rebooking a move is a feature only for the `reject` event

File: app/services/move_events/params_validator.rb (hunk 4 above)
Comment (martyn-w) [reply app/services/move_events/params_validator.rb:11]: sorry that wasn't clear - I'll add a more detailed note to explain the situation

File: spec/requests/api/move_events_controller_lockout_spec.rb (hunk 5 above)
Comment (martyn-w) [reply spec/requests/api/move_events_controller_lockout_spec.rb:142]: ok, I have moved the missing location to the params validator spec

File: spec/requests/api/move_events_controller_redirect_spec.rb (hunk 6 above)
Comment (martyn-w) [reply spec/requests/api/move_events_controller_redirect_spec.rb:116]: I have removed a validation test and am checking in the params validator instead

**Pull request merged**